livre/core/cache/Cache.cpp

```diff
@@ -28,125 +27,263 @@
+    }
 
-    for( CacheMap::const_iterator it = _cacheMap.begin(); it != _cacheMap.end(); ++it )
+    bool isSatisfied( const Cache& cache ) const
     {
-        const CacheObjectPtr& object = it->second;
-        if( object && object->isValid() && object->_isLoaded() )
-            cacheObjects.push_back( object.get( ));
+        const size_t usedMemBytes = cache.getStatistics().getUsedMemory();
+        return usedMemBytes < ( 1.0f - _cleanUpRatio ) * _maxMemBytes;
     }
 
-    if( cacheObjects.empty( ))
-        return AR_EMPTY;
+    void insert( const CacheId& cacheId )
+    {
+        remove( cacheId );
+        _lruQueue.push_back( cacheId );
+    }
 
-    std::vector< CacheObject * > modifiedList;
-    cachePolicy._apply( *this, cacheObjects, modifiedList );
+    void remove( const CacheId& cacheId )
+    {
+        typename LRUQueue::iterator it = _lruQueue.begin();
+        while( it != _lruQueue.end( ))
+        {
+            if( *it == cacheId )
+                _lruQueue.erase( it );
+            else
+                ++it;
+        }
+    }
 
-    _unload( cachePolicy, modifiedList );
-    return AR_ACTIVATED;
-}
+    CacheIds getObjects() const
+    {
+        CacheIds ids;
+        ids.reserve( _lruQueue.size( ));
+        ids.insert( ids.begin(), _lruQueue.begin(), _lruQueue.end());
+        return ids;
+    }
 
-Cache::Cache()
-    : _statistics( new CacheStatistics( "Statistics", CACHE_LOG_SIZE ) )
-{
-}
+    const size_t _maxMemBytes;
+    const float _cleanUpRatio;
+    LRUQueue _lruQueue;
+};
 
-Cache::~Cache()
+struct Cache::Impl
 {
-    for( CacheMap::iterator it = _cacheMap.begin(); it != _cacheMap.end(); ++it )
+    Impl( Cache& cache,
+          const size_t maxMemBytes )
+        : _policy( maxMemBytes )
+        , _cache( cache )
+        , _statistics( "Statistics", CACHE_LOG_SIZE )
     {
-        CacheObjectPtr& cacheObject = it->second;
-        cacheObject->_unregisterObserver( this );
-        cacheObject->_unregisterObserver( _statistics.get( ));
     }
-}
 
-CacheObjectPtr Cache::_get( const CacheId& cacheId )
-{
-    if( cacheId == INVALID_CACHE_ID )
-        return CacheObjectPtr();
+    ~Impl()
+    {}
 
-    WriteLock writeLock( _mutex );
-    CacheMap::iterator it = _cacheMap.find( cacheId );
-    if( it == _cacheMap.end() )
+    void applyPolicy() const
     {
-        CacheObjectPtr cacheObject( _generate( cacheId ));
-        cacheObject->_registerObserver( this );
-        cacheObject->_registerObserver( _statistics.get() );
-        _cacheMap[ cacheId ] = cacheObject;
+        if( _cacheMap.empty() || !_policy.isActivated( _cache ) )
+            return;
+
+        unload();
     }
 
-    LBASSERT( _cacheMap[ cacheId ]->_status );
+    void onDelete( CacheObject* cacheObject ) const
+    {
+        WriteLock lock( _mutex );
+        cacheObject->_decreaseRef();
+    }
 
-    return _cacheMap[ cacheId ];
-}
+    CacheObjectPtr getSharedPtr( CacheObject* obj ) const
```

Comment:
Keeping a shared_ptr in CacheMap is not feasible. Because if you are at the state of unloading after checking use count and meanwhile somebody has copying of shared_ptr you should not be deleting the data. Therefore unload and ref count should be protected by the same lock. You cant access internal lock of a shared_ptr.